packages/backend/src/sagas/postActivitySaga.js

```diff
@@ -0,0 +1,111 @@
+import {
+  cancelled,
+  fork,
+  join,
+  put,
+  select,
+  take
+} from 'redux-saga/effects';
+
+import callWithin from './effects/callWithin';
+import observeOnce from './effects/observeOnce';
+import whileConnected from './effects/whileConnected';
+
+import deleteKey from '../util/deleteKey';
+import getTimestamp from '../util/getTimestamp';
+import uniqueID from '../util/uniqueID';
+
+import {
+  POST_ACTIVITY,
+  POST_ACTIVITY_FULFILLED,
+  POST_ACTIVITY_PENDING,
+  POST_ACTIVITY_REJECTED
+} from '../Actions/postActivity';
+
+import { UPSERT_ACTIVITY } from '../Actions/upsertActivity';
+
+const SEND_TIMEOUT = 5000;
+
+export default function* () {
+  yield whileConnected(function* (directLine, userID) {
+    for (let numActivitiesPosted = 0;; numActivitiesPosted++) {
+      const action = yield take(POST_ACTIVITY);
+
+      yield fork(postActivity, directLine, userID, numActivitiesPosted, action);
+    }
+  });
+}
+
+function* postActivity(directLine, userID, numActivitiesPosted, { payload: { activity } }) {
+  const locale = yield select(({ settings: { language } }) => language);
+  const { attachments, channelData: { clientActivityID = uniqueID() } = {} } = activity;
+
+  activity = {
+    ...deleteKey(activity, 'id'),
+    attachments: attachments && attachments.map(({ contentType, contentUrl, name }) => ({
+      contentType,
+      contentUrl,
+      name
+    })),
+    channelData: {
+      clientActivityID,
+      ...deleteKey(activity.channelData, 'state')
+    },
+    channelId: 'webchat',
+    from: {
+      id: userID,
+      role: 'user'
+    },
+    locale,
+    timestamp: getTimestamp()
+  };
+
+  if (!numActivitiesPosted) {
+    activity.entities = [...activity.entities || [], {
+      // TODO: Clean up the capabilities
+      requiresBotState: true,
+      supportsListening: true,
+      supportsTts: true,
+      type: 'ClientCapabilities'
+    }];
+  }
+
+  const meta = { clientActivityID };
+
+  yield put({ type: POST_ACTIVITY_PENDING, payload: { activity }, meta });
+
+  try {
+    // Quirks: We might receive UPSERT_ACTIVITY before the postActivity call completed
+    //         So, we setup expectation first, then postActivity afterward
```

Comment:
might get message back before it is sent... is this for local bots that reply inline?

Reply:
E.g. WebSocket use a different channel. If WS connection is fast enough, it will echo back the user activity before the PostActivity REST API complete.